File: 28a-distributed-systems-deep-dive/applied/idempotency.py
# Companion code for "The Backend of Luck" - Chapter 28a, Distributed Systems Deep Dive.
# https://thebackendofluck.com | https://github.com/thebackendofluck/book
# SPDX-License-Identifier: Apache-2.0
#
# FOR TESTING AND EVALUATION ONLY. NOT FOR PRODUCTION USE.
# Published to demonstrate the patterns explained in the book. This code is
# not certified for real-money gaming: operating a gambling platform requires
# your own licence, independent test-lab certification (GLI, eCOGRA or
# equivalent) and regulator approval.

"""
FastAPI Idempotency-Key middleware.

Algorithm (corrected to address "first-wins with in-progress state" gap):

1. On a write request (POST/PUT/PATCH/DELETE):
   a. If the path matches the REQUIRED whitelist and no Idempotency-Key header is
      present, reject with 400.
   b. If the header is absent and the path is not required, pass through.

2. Canonicalize the body (SHA-256 over JSON with sorted keys; raw bytes if not JSON).

3. Attempt to acquire the key via INSERT-as-lock:

       INSERT INTO idempotency_records (key, user_id, path, body_hash, state, ...)
       VALUES (..., 'in_progress')
       ON CONFLICT (key) DO NOTHING
       RETURNING key

   - If RETURNING returned the key, this request is the OWNER. Execute the handler.
     When complete and the response is terminal (2xx final, 4xx client-error, 5xx
     final), persist: state='terminal', response_status, response_body. If the
     response is non-terminal (202 Accepted, 102 Processing), do NOT cache; delete
     the in_progress record so a retry can be a new owner.

   - If RETURNING returned nothing, another request won. Fetch current record:
       * If body_hash differs -> 409 Conflict (key reuse).
       * If state='terminal' -> replay cached response_status + response_body.
       * If state='in_progress' -> poll with exponential backoff until either
         terminal or TIMEOUT (default 30s). On timeout -> 409 Conflict
         ("original request still processing").

4. TTLs:
   - wallet/payment endpoints: 7 days
   - everything else: 24 hours

A background task (pg_cron or k8s CronJob) cleans expired rows daily; this module
provides `purge_expired()` for the cron wrapper.

This implementation is intentionally datastore-agnostic: it accepts a callable
`store` that implements the IdempotencyStore protocol, so tests can use a
fake in-memory store and prod uses a psycopg-backed implementation.
"""

from __future__ import annotations

import asyncio
import hashlib
import json
import logging
import time
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Iterable, Optional, Protocol

from fastapi import Request, Response
from fastapi.responses import JSONResponse
from pydantic import Field
from pydantic_settings import BaseSettings
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import ASGIApp

logger = logging.getLogger(__name__)


STATE_IN_PROGRESS = "in_progress"
STATE_TERMINAL = "terminal"


# ---------------------------------------------------------------------------
# Settings
# ---------------------------------------------------------------------------


class IdempotencySettings(BaseSettings):
    """Configuration for the idempotency middleware."""

    enabled: bool = True

    required_path_prefixes: tuple[str, ...] = (
        "/wallet/",
        "/gal/",
        "/payments/",
        "/kyc/",
        "/bonus/",
        "/responsible-gaming/self-exclusion",
        "/dsr/",
        "/consent/record",
    )

    long_ttl_path_prefixes: tuple[str, ...] = (
        "/wallet/",
        "/payments/",
    )

    long_ttl_seconds: int = 60 * 60 * 24 * 7
    default_ttl_seconds: int = 60 * 60 * 24

    inflight_poll_interval_ms: int = 100
    inflight_poll_timeout_seconds: int = 30

    max_key_length: int = 255
    max_cached_body_bytes: int = 1024 * 1024

    idem_methods: frozenset[str] = frozenset({"POST", "PUT", "PATCH", "DELETE"})

    model_config = {"env_prefix": "IDEMPOTENCY_"}


# ---------------------------------------------------------------------------
# Records & Store protocol
# ---------------------------------------------------------------------------


@dataclass
class IdempotencyRecord:
    key: str
    user_id: Optional[str]
    path: str
    body_hash: str
    state: str
    response_status: Optional[int] = None
    response_body: Optional[bytes] = None
    response_headers: dict[str, str] = field(default_factory=dict)
    created_at: float = 0.0
    expires_at: float = 0.0


class IdempotencyStore(Protocol):
    """Abstract store. Implementations must be atomic on try_acquire."""

    async def try_acquire(self, record: IdempotencyRecord) -> bool:
        """
        Insert in_progress record atomically. Return True if this caller is the
        owner; False if another caller already holds the key.
        """

    async def fetch(self, key: str) -> Optional[IdempotencyRecord]:
        ...

    async def mark_terminal(
        self,
        key: str,
        status: int,
        body: bytes,
        headers: dict[str, str],
    ) -> None:
        ...

    async def release_inflight(self, key: str) -> None:
        """Delete an in_progress row (used when the response is non-terminal)."""

    async def purge_expired(self, now: float) -> int:
        ...


# ---------------------------------------------------------------------------
# Hashing helpers
# ---------------------------------------------------------------------------


def canonical_body_hash(body: bytes) -> str:
    """SHA-256 over canonicalized body.

    - If body is valid JSON: re-serialize with sorted keys + no whitespace.
    - Otherwise: hash raw bytes.
    """
    if not body:
        return hashlib.sha256(b"").hexdigest()
    try:
        parsed = json.loads(body)
    except (ValueError, TypeError):
        return hashlib.sha256(body).hexdigest()
    canon = json.dumps(parsed, sort_keys=True, separators=(",", ":")).encode()
    return hashlib.sha256(canon).hexdigest()


def is_terminal_status(status: int) -> bool:
    """
    A response is terminal (cacheable) if it represents a final resolution of the
    request. 202 Accepted and 102 Processing are NOT terminal.
    """
    if status == 202 or status == 102:
        return False
    return 200 <= status < 600


# ---------------------------------------------------------------------------
# Middleware
# ---------------------------------------------------------------------------


UserExtractor = Callable[[Request], Optional[str]]


class IdempotencyMiddleware(BaseHTTPMiddleware):
    def __init__(
        self,
        app: ASGIApp,
        store: IdempotencyStore,
        settings: Optional[IdempotencySettings] = None,
        user_id_extractor: Optional[UserExtractor] = None,
    ) -> None:
        super().__init__(app)
        self.store = store
        self.settings = settings or IdempotencySettings()
        self.user_id_extractor = user_id_extractor or (
            lambda r: r.headers.get("x-user-id")
        )

    def _ttl_for(self, path: str) -> int:
        for prefix in self.settings.long_ttl_path_prefixes:
            if path.startswith(prefix):
                return self.settings.long_ttl_seconds
        return self.settings.default_ttl_seconds

    def _path_requires_key(self, path: str) -> bool:
        return any(
            path.startswith(p) for p in self.settings.required_path_prefixes
        )

    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        if not self.settings.enabled:
            return await call_next(request)

        if request.method not in self.settings.idem_methods:
            return await call_next(request)

        key = request.headers.get("idempotency-key")
        path = request.url.path

        if not key:
            if self._path_requires_key(path):
                return JSONResponse(
                    {
                        "error": "idempotency_key_required",
                        "message": (
                            "Idempotency-Key header is required for this endpoint."
                        ),
                    },
                    status_code=400,
                )
            return await call_next(request)

        if len(key) > self.settings.max_key_length:
            return JSONResponse(
                {"error": "idempotency_key_too_long"}, status_code=400
            )

        body = await request.body()
        body_hash = canonical_body_hash(body)
        user_id = self.user_id_extractor(request)

        now = time.time()
        record = IdempotencyRecord(
            key=key,
            user_id=user_id,
            path=path,
            body_hash=body_hash,
            state=STATE_IN_PROGRESS,
            created_at=now,
            expires_at=now + self._ttl_for(path),
        )

        acquired = await self.store.try_acquire(record)

        if not acquired:
            return await self._handle_non_owner(key, body_hash)

        # We own the key. Rebuild the request so the handler can still read body.
        async def _receive() -> dict[str, Any]:
            return {"type": "http.request", "body": body, "more_body": False}

        request._receive = _receive  # type: ignore[attr-defined]

        try:
            response = await call_next(request)
        except Exception:
            await self.store.release_inflight(key)
            raise

        # Capture response body
        chunks: list[bytes] = []
        async for chunk in response.body_iterator:  # type: ignore[attr-defined]
            chunks.append(chunk)
        payload = b"".join(chunks)

        status = response.status_code
        headers = {k: v for k, v in response.headers.items()}

        if is_terminal_status(status) and len(payload) <= self.settings.max_cached_body_bytes:
            await self.store.mark_terminal(key, status, payload, headers)
        else:
            # Non-terminal (202 pending, etc.) or body too large: do not cache.
            await self.store.release_inflight(key)

        return Response(
            content=payload,
            status_code=status,
            headers=headers,
            media_type=response.media_type,
        )

    async def _handle_non_owner(self, key: str, body_hash: str) -> Response:
        """Another request owns the key. Validate body and poll for terminal."""
        existing = await self.store.fetch(key)
        if existing is None:
            # Race: owner released between try_acquire and fetch. Safest: 409.
            return JSONResponse(
                {"error": "idempotency_race_retry"}, status_code=409
            )
        if existing.body_hash != body_hash:
            return JSONResponse(
                {
                    "error": "idempotency_key_reused_with_different_body",
                    "message": (
                        "Idempotency-Key was previously used with a different "
                        "request body."
                    ),
                },
                status_code=409,
            )

        if existing.state == STATE_TERMINAL:
            return self._replay(existing)

        # Poll
        deadline = time.time() + self.settings.inflight_poll_timeout_seconds
        interval = self.settings.inflight_poll_interval_ms / 1000.0
        while time.time() < deadline:
            await asyncio.sleep(interval)
            current = await self.store.fetch(key)
            if current is None:
                # Owner released (non-terminal path). Tell client to retry.
                return JSONResponse(
                    {"error": "idempotency_owner_released_retry"}, status_code=409
                )
            if current.state == STATE_TERMINAL:
                return self._replay(current)
            interval = min(interval * 1.5, 1.0)

        return JSONResponse(
            {
                "error": "idempotency_original_request_timeout",
                "message": "Original request still processing after timeout.",
            },
            status_code=409,
        )

    def _replay(self, rec: IdempotencyRecord) -> Response:
        assert rec.response_status is not None
        return Response(
            content=rec.response_body or b"",
            status_code=rec.response_status,
            headers={**rec.response_headers, "x-idempotent-replay": "true"},
        )

File: 44-deploying-igaming-on-cloudflare-workers/applied/idempotency.ts
// Companion code for "The Backend of Luck" - Chapter 44, Deploying iGaming Platforms on Cloudflare Workers.
// https://thebackendofluck.com | https://github.com/thebackendofluck/book
// SPDX-License-Identifier: Apache-2.0
//
// FOR TESTING AND EVALUATION ONLY. NOT FOR PRODUCTION USE.
// Published to demonstrate the patterns explained in the book. This code is
// not certified for real-money gaming: operating a gambling platform requires
// your own licence, independent test-lab certification (GLI, eCOGRA or
// equivalent) and regulator approval.

/**
 * D1-backed idempotency helper for Cloudflare Workers.
 *
 * INSERT-as-lock pattern using SQLite's ON CONFLICT DO NOTHING. If the insert
 * wins, the caller executes the handler and then UPDATEs the row to terminal.
 * If the insert loses, the caller fetches the row and either replays the
 * cached response or rejects (409) on body-hash mismatch or in-progress timeout.
 *
 * Non-terminal responses (202 Accepted) are NOT cached; the in-progress row
 * is deleted so a retry can re-execute.
 */

export interface IdempotencyRecord {
  key: string;
  user_id: string | null;
  path: string;
  body_hash: string;
  state: "in_progress" | "terminal";
  response_status: number | null;
  response_body: string | null;
  response_headers: string | null;
  created_at: number;
  expires_at: number;
}

export interface IdempotencyBinding {
  DB: D1Database;
}

const LONG_TTL_MS = 7 * 24 * 60 * 60 * 1000;
const DEFAULT_TTL_MS = 24 * 60 * 60 * 1000;
const POLL_INTERVAL_MS = 100;
const POLL_TIMEOUT_MS = 30_000;

const LONG_TTL_PREFIXES = ["/api/payments/", "/api/wallet/"];

const REQUIRED_PREFIXES = [
  "/api/payments/deposit",
  "/api/payments/withdraw",
  "/api/wallet/",
];

export function pathRequiresKey(path: string): boolean {
  return REQUIRED_PREFIXES.some((p) => path.startsWith(p));
}

function ttlForPath(path: string): number {
  return LONG_TTL_PREFIXES.some((p) => path.startsWith(p))
    ? LONG_TTL_MS
    : DEFAULT_TTL_MS;
}

export function isTerminalStatus(status: number): boolean {
  if (status === 102 || status === 202) return false;
  return status >= 200 && status < 600;
}

export async function canonicalBodyHash(body: string): Promise<string> {
  let canonical = body;
  try {
    const parsed = JSON.parse(body);
    canonical = canonicalStringify(parsed);
  } catch {
    /* fall through, hash raw */
  }
  const buf = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder().encode(canonical),
  );
  return [...new Uint8Array(buf)]
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

function canonicalStringify(v: unknown): string {
  if (v === null || typeof v !== "object") return JSON.stringify(v);
  if (Array.isArray(v)) {
    return "[" + v.map(canonicalStringify).join(",") + "]";
  }
  const obj = v as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  return (
    "{" +
    keys
      .map((k) => JSON.stringify(k) + ":" + canonicalStringify(obj[k]))
      .join(",") +
    "}"
  );
}

export async function tryAcquire(
  db: D1Database,
  rec: Omit<IdempotencyRecord, "response_status" | "response_body" | "response_headers">,
): Promise<boolean> {
  const result = await db
    .prepare(
      `INSERT INTO idempotency_records
         (key, user_id, path, body_hash, state, created_at, expires_at)
       VALUES (?, ?, ?, ?, 'in_progress', ?, ?)
       ON CONFLICT (key) DO NOTHING
       RETURNING key`,
    )
    .bind(
      rec.key,
      rec.user_id,
      rec.path,
      rec.body_hash,
      rec.created_at,
      rec.expires_at,
    )
    .first<{ key: string }>();
  return result !== null;
}

export async function fetchRecord(
  db: D1Database,
  key: string,
): Promise<IdempotencyRecord | null> {
  const now = Date.now();
  const row = await db
    .prepare(
      `SELECT * FROM idempotency_records WHERE key = ? AND expires_at > ?`,
    )
    .bind(key, now)
    .first<IdempotencyRecord>();
  return row ?? null;
}

export async function markTerminal(
  db: D1Database,
  key: string,
  status: number,
  body: string,
  headers: Record<string, string>,
): Promise<void> {
  await db
    .prepare(
      `UPDATE idempotency_records
       SET state = 'terminal',
           response_status = ?,
           response_body = ?,
           response_headers = ?
       WHERE key = ?`,
    )
    .bind(status, body, JSON.stringify(headers), key)
    .run();
}

export async function releaseInflight(
  db: D1Database,
  key: string,
): Promise<void> {
  await db
    .prepare(
      `DELETE FROM idempotency_records WHERE key = ? AND state = 'in_progress'`,
    )
    .bind(key)
    .run();
}

/**
 * Wraps a handler with idempotency semantics.
 *
 *   return withIdempotency(request, env.DB, userId, () => realHandler(request));
 */
export async function withIdempotency(
  request: Request,
  db: D1Database,
  userId: string | null,
  handler: (body: string) => Promise<Response>,
): Promise<Response> {
  if (!["POST", "PUT", "PATCH", "DELETE"].includes(request.method)) {
    return handler(await request.text());
  }

  const url = new URL(request.url);
  const key = request.headers.get("idempotency-key");
  const body = await request.text();

  if (!key) {
    if (pathRequiresKey(url.pathname)) {
      return json(
        { error: "idempotency_key_required" },
        400,
      );
    }
    return handler(body);
  }

  if (key.length > 255) {
    return json({ error: "idempotency_key_too_long" }, 400);
  }

  const bodyHash = await canonicalBodyHash(body);
  const now = Date.now();
  const expiresAt = now + ttlForPath(url.pathname);

  const acquired = await tryAcquire(db, {
    key,
    user_id: userId,
    path: url.pathname,
    body_hash: bodyHash,
    state: "in_progress",
    created_at: now,
    expires_at: expiresAt,
  });

  if (!acquired) {
    return handleNonOwner(db, key, bodyHash);
  }

  let response: Response;
  try {
    response = await handler(body);
  } catch (err) {
    await releaseInflight(db, key);
    throw err;
  }

  const respBody = await response.clone().text();
  const headers: Record<string, string> = {};
  response.headers.forEach((v, k) => {
    headers[k] = v;
  });

  if (isTerminalStatus(response.status) && respBody.length <= 1_048_576) {
    await markTerminal(db, key, response.status, respBody, headers);
  } else {
    await releaseInflight(db, key);
  }

  return response;
}

async function handleNonOwner(
  db: D1Database,
  key: string,
  bodyHash: string,
): Promise<Response> {
  const existing = await fetchRecord(db, key);
  if (!existing) {
    return json({ error: "idempotency_race_retry" }, 409);
  }
  if (existing.body_hash !== bodyHash) {
    return json(
      { error: "idempotency_key_reused_with_different_body" },
      409,
    );
  }
  if (existing.state === "terminal") {
    return replay(existing);
  }

  const deadline = Date.now() + POLL_TIMEOUT_MS;
  let interval = POLL_INTERVAL_MS;
  while (Date.now() < deadline) {
    await sleep(interval);
    const cur = await fetchRecord(db, key);
    if (!cur) {
      return json({ error: "idempotency_owner_released_retry" }, 409);
    }
    if (cur.state === "terminal") return replay(cur);
    interval = Math.min(interval * 1.5, 1000);
  }
  return json({ error: "idempotency_original_request_timeout" }, 409);
}

function replay(rec: IdempotencyRecord): Response {
  const headers = rec.response_headers
    ? (JSON.parse(rec.response_headers) as Record<string, string>)
    : {};
  headers["x-idempotent-replay"] = "true";
  return new Response(rec.response_body ?? "", {
    status: rec.response_status ?? 200,
    headers,
  });
}

function json(body: unknown, status: number): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "content-type": "application/json" },
  });
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
